DeviceDetector.php

```diff
@@ -191,105 +191,116 @@ public function __construct($userAgent = '')
+        foreach (AbstractDeviceParser::getAvailableDeviceTypes() as $deviceName => $deviceType) {
+            if (\strtolower($methodName) === 'is' . \strtolower(\str_replace(' ', '', $deviceName))) {
+                return $this->getDevice() === $deviceType;
             }
         }
 
         foreach (self::$clientTypes as $client) {
-            if (strtolower($methodName) == 'is' . strtolower(str_replace(' ', '', $client))) {
-                return $this->getClient('type') == $client;
+            if (\strtolower($methodName) === 'is' . \strtolower(\str_replace(' ', '', $client))) {
+                return $this->getClient('type') === $client;
             }
         }
 
-        throw new \BadMethodCallException("Method $methodName not found");
+        throw new \BadMethodCallException("Method {$methodName} not found");
     }
 
     /**
      * Sets the useragent to be parsed
      *
      * @param string $userAgent
      */
-    public function setUserAgent($userAgent)
+    public function setUserAgent(string $userAgent): void
     {
-        if ($this->userAgent != $userAgent) {
+        if ($this->userAgent !== $userAgent) {
             $this->reset();
         }
-        $this->userAgent = $userAgent;
-    }
 
-    protected function reset()
-    {
-        $this->bot = null;
-        $this->client = null;
-        $this->device = null;
-        $this->os = null;
-        $this->brand = '';
-        $this->model = '';
-        $this->parsed = false;
+        $this->userAgent = $userAgent;
     }
 
     /**
-     * @param ClientParserAbstract|string $parser
+     * @param AbstractClientParser|string $parser
+     *
      * @throws \Exception
      */
-    public function addClientParser($parser)
+    public function addClientParser($parser): void
     {
-        if (is_string($parser) && class_exists('DeviceDetector\\Parser\\Client\\' . $parser)) {
-            $className = 'DeviceDetector\\Parser\\Client\\' . $parser;
+        $className = 'DeviceDetector\\Parser\\Client\\' . $parser;
+
+        if (\is_string($parser) && \class_exists($className)) {
             $parser = new $className();
         }
 
-        if ($parser instanceof ClientParserAbstract) {
+        if ($parser instanceof AbstractClientParser) {
             $this->clientParsers[] = $parser;
-            self::$clientTypes[] = $parser->getName();
+            self::$clientTypes[]   = $parser->getName();
+
             return;
         }
 
         throw new \Exception('client parser not found');
     }
 
-    public function getClientParsers()
+    /**
+     * @return array<AbstractClientParser>
+     */
+    public function getClientParsers(): array
     {
         return $this->clientParsers;
     }
 
     /**
-     * @param DeviceParserAbstract|string $parser
+     * @param AbstractDeviceParser|string $parser
+     *
      * @throws \Exception
      */
-    public function addDeviceParser($parser)
+    public function addDeviceParser($parser): void
```

Comment:
Improvement:
```php
public function addDeviceParser(string $parser): void {
    if (\is_string($parser) && \class_exists($className)) {
            $parser = new $className();
    }
    if ($parser instanceof AbstractDeviceParser) {
            $this->deviceParsers[] = $parser;
            return;
        }
}

// And add classes like this
$deviceDetector = new DeviceDetector($ua);
$deviceDetector->addDeviceParser(Mobile::class);
$deviceDetector->addDeviceParser(Notebook::class);
```

Third-party classes of our own production will also work here
```
  class Fridge extends AbstractDeviceParser 
  {
  } 
```

Reply:
sounds fine

Reply:
on the other side, we could directly pass instances and remove all the checks for valid classes...

Reply:
Also a good option.
```php

public function addDeviceParser(AbstractDeviceParser $parser): void {
    if ($parser instanceof AbstractDeviceParser) {
            $this->deviceParsers[] = $parser;
            return;
   }
     throw new \Exception('client parser not found');
}

```